de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/editor/ServerEditorManager.java

```diff
@@ -0,0 +1,127 @@
+import de.fu_berlin.inf.dpp.activities.TextEditActivity;
+import de.fu_berlin.inf.dpp.editor.IEditorManager;
+import de.fu_berlin.inf.dpp.editor.ISharedEditorListener;
+import de.fu_berlin.inf.dpp.editor.text.LineRange;
+import de.fu_berlin.inf.dpp.editor.text.TextSelection;
+import de.fu_berlin.inf.dpp.filesystem.IProject;
+import de.fu_berlin.inf.dpp.session.User;
+import org.apache.commons.collections4.map.LRUMap;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ServerEditorManager implements IEditorManager {
+
+    private static final Logger LOG = Logger
+        .getLogger(ServerEditorManager.class);
+
+    private Map<SPath, Editor> openEditors = Collections.synchronizedMap(new LRUMap(10));
+    private List<ISharedEditorListener> listeners = new CopyOnWriteArrayList();
+
+    @Override public void openEditor(SPath path, boolean activate) {
+        try {
+            getOrCreateEditor(path);
+        } catch (IOException e) {
+            LOG.warn("Could not open editor for "+path);
+        }
+    }
+
+    @Override public Set<SPath> getOpenEditors() {
+        return openEditors.keySet();
+    }
+
+    @Override public String getContent(SPath path) {
+        try {
+            return getOrCreateEditor(path).getContent();
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @Override public void saveEditors(IProject project) {
+        // do nothing?
+        // we do not keep dirty editors,
+        // because the LRUMap might close Editors at any time
+    }
+
+    @Override public void closeEditor(SPath path) {
+        openEditors.remove(path);
+    }
+
+    @Override public void adjustViewport(SPath path, LineRange range,
+        TextSelection selection) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override public void jumpToUser(User target) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override public void addSharedEditorListener(
+        ISharedEditorListener listener) {
+        listeners.add(listener);
+    }
+
+    @Override public void removeSharedEditorListener(
+        ISharedEditorListener listener) {
+        listeners.remove(listener);
+    }
+
+    private Editor getOrCreateEditor(SPath path) throws IOException {
+        Editor editor = openEditors.get(path);
+        if (editor == null) {
+            editor = new Editor(path);
+            openEditors.put(path, editor);
+        }
+        return editor;
+    }
+
+    /**
+     * Executes a text edit activity on the matching editor.
+     *
+     * @param activity
+     *            the activity describing the text edit to apply
+     */
+    public void applyTextEdit(TextEditActivity activity) {
+        SPath path = activity.getPath();
+        try {
+            Editor editor = getOrCreateEditor(path);
+            editor.applyTextEdit(activity);
+            editor.save();
```

Comment:
It seems a bit wasteful to implement an editor that caches the changes to the file and then to still save them to disk after every `TextEditActivity`. It would be better to just save the changes to disk when needed (for example when closing the editor; I would also save the changes before moving/renaming a file, but this might not be strictly necessary).

From your comment in `saveEditor(SPath)`, I get that you don't know when an editor is removed from the `LRUMap`.

This problem can be avoided by removing the least recently used editor yourself before adding a new one when the map is full. Even though the `LRUMap` does not offer such a method itself (that removes *and returns* the LRU object), you can implement this logic yourself:

Before adding a new editor in `getOrCreateEditor(SPath)`, you can add a check whether the map is full. If it is, you can remove the LRU entry by using an [`OrderedMapIterator`](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/OrderedMapIterator.html), which is ordered from least to most recently used entry [according to the `LRUMap` doc](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/map/LRUMap.html).
```Java
if (openEditors.isFull) {
    MapIterator mapIterator = openEditors.mapIterator();
    mapIterator.next();
    Editor editor = mapIterator.getValue();
    mapIterator.remove();
    editor.save();
}
```
(Since a `LRUMap` has a minimum capacity of one, a check to `Iterator.hasNext()` should not be necessary for the first `next()` call when the map is full.)

You could shorten this check a bit by using `closeEditor(SPath)`, but I am not sure whether this would create a `ConcurrentModificationException` with the Iterator, even though the iterator is no longer used:
```Java
if (openEditors.isFull) {
    MapIterator mapIterator = openEditors.mapIterator();
    mapIterator.next();
    closeEditor(mapIterator.getValue().getFilePath());
}
```
This works under the assumption that `closeEditor(SPath)` now saves the closed editor.

Either way, I would not rely on the `LRUMap` removing the correct enty itself when a new entry is added as I am worried us getting the LRU entry could change the order, meaning the automatically removed LRU entry could be a different entry we have not saved.

But all of this is just based on reading the `LRUMap` doc. I haven't actually tested it, so it might not work correctly or create other problems. Also, I am not sure whether this optimization is worth it.

What is your opinion?

Reply:
Alright I *think* I have solved all other outstanding issues, but please check, if the new commit does satisfy you. I appreciate the feedback especially on docs and naming, I always forget that, especially when porting older patchsets, like this one.

 The currently unused methods will be used by a future pr (specifically this one, which I will update soon-ish: https://github.com/saros-project/saros/pull/147 )

I agree with your suggestion and will most likely implement it, although I am neither sure, if it is really necessary from a performance stand point, nor happy about it from a code stand point, as it looks like a hack to do something the class was designed to handle for us manually.

However I will have to test it again, so I will get back to it later.